mds/models.py

```diff
@@ -109,21 +109,16 @@ def __str__(self):
 
 
 class DeviceQueryset(models.QuerySet):
-    def with_latest_event(self):
-        return self.prefetch_related(
-            Prefetch(
-                "event_records",
-                queryset=EventRecord.objects.filter(
-                    id__in=Subquery(
-                        EventRecord.objects.filter(device_id=OuterRef("device_id"))
-                        .exclude(event_type="telemetry")
-                        .order_by("-timestamp")
-                        .values_list("id", flat=True)[:1]
-                    )
-                ),
-                to_attr="_latest_event",
-            )
+    def with_latest_events(self):
+        prefetchedEvents = Prefetch(
+            "event_records",
+            queryset=EventRecord.objects.exclude(event_type="telemetry").order_by(
+                "-timestamp"
+            ),
+            to_attr="_latest_events",
         )
+        prefetch_related = self.prefetch_related(prefetchedEvents)
```

Comment:
Are your still filtering by this device's ID?

Reply:
Maybe I am wrong, but I think the way it works is that:
- when you have a single Device object d, if you do d.latest_event, the latest_event property will do 1 query on EventRecord, filtering via d.id (so no prefetch actions)
- when you are listing devices, such as Device.objects.all(), if you want latest_event property in each device, you have to do Device.objects.with_latest_events.all().  In such a case, the prefetch in with_latest_events will do 1 single big query taking all eventrecords,  and python in the background will do the association for which eventRecord correspond to which device. So there is no need to filter via ID.